core/src/test/java/io/questdb/test/preferences/PreferencesStoreConcurrentTest.java

```diff
@@ -0,0 +1,119 @@
+        final int numOfReads = 500;
+        assertMemoryLeak(() -> {
+            final String preferences = "{\"key1\":\"value1\",\"key2\":\"value2\",\"key3\":\"value3\"}";
+            try (PreferencesStore preferencesStore = new PreferencesStore(engine.getConfiguration())) {
+                preferencesStore.init();
+                try (DirectUtf8Sink directUtf8Sink = new DirectUtf8Sink(256)) {
+                    for (int j = 0; j < numOfWrites; j++) {
+                        directUtf8Sink.clear();
+                        directUtf8Sink.put(preferences);
+                        preferencesStore.save(directUtf8Sink, PreferencesStore.Mode.OVERWRITE, j);
+                    }
+                }
+
+                final ConcurrentHashMap<Integer, Throwable> errors = new ConcurrentHashMap<>();
+
+                final CyclicBarrier start = new CyclicBarrier(numOfWriterThreads + numOfReaderThreads);
+                final SOCountDownLatch end = new SOCountDownLatch(numOfWriterThreads + numOfReaderThreads);
+                for (int i = 0; i < numOfWriterThreads; i++) {
+                    final int threadIndex = i;
+                    new Thread(() -> {
+                        await(start);
+                        try (DirectUtf8Sink directUtf8Sink = new DirectUtf8Sink(256)) {
+                            for (int j = 0; j < numOfWrites; j++) {
+                                directUtf8Sink.clear();
+                                directUtf8Sink.put(preferences);
+                                try {
+                                    preferencesStore.save(directUtf8Sink, PreferencesStore.Mode.OVERWRITE, preferencesStore.getVersion());
+                                } catch (CairoException e) {
+                                    if (!e.isPreferencesOutOfDateError()) {
+                                        throw e;
+                                    }
+                                }
+                            }
+                        } catch (Throwable th) {
+                            errors.put(threadIndex, th);
+                        }
+                        end.countDown();
+                    }).start();
+                }
+
+                for (int i = 0; i < numOfReaderThreads; i++) {
+                    final int threadIndex = i;
+                    new Thread(() -> {
+                        await(start);
+                        try {
+                            final Utf8StringSink sink = new Utf8StringSink();
+                            for (int j = 0; j < numOfReads; j++) {
+                                sink.clear();
+                                preferencesStore.populateSettings(sink);
+                                assertEquals("\"preferences\":{\"key1\":\"value1\",\"key2\":\"value2\",\"key3\":\"value3\"}", sink);
+                            }
+                        } catch (Throwable th) {
+                            errors.put(threadIndex, th);
```

Comment:
reader and writer threads will be overwriting each other errors

Reply:
fixed in https://github.com/questdb/questdb/pull/5621/commits/afd5fc4cefb807a582a6b827c407955587f7cdef